build.rs

```diff
@@ -0,0 +1,211 @@
+use std::collections::{HashMap, HashSet};
+use std::env;
+use std::fs::{self, File};
+use std::io::Write;
+use std::path::Path;
+
+use serde::Deserialize;
+
+#[derive(Deserialize, Debug)]
+pub struct ModelData {
+    pub norm: FieldsDescribe,
+    pub weights: Weights,
+    pub fields: Vec<String>,
+    pub alpha: f32,
+}
+
+impl ModelData {
+    fn into_src(&self, model_name: &str) -> (HashSet<usize>, String) {
+        let fields = self
+            .fields
+            .iter()
+            .fold(String::new(), |acc, f| acc + "\"" + f + "\".to_string(), ");
+
+        let i_size = self.fields.len();
+        let l0_size = self.weights.l0_bias.len();
+        let l1_size = self.weights.l1_bias.len();
+        let o_size = self.weights.l2_bias.len();
+
+        if l0_size != l1_size {
+            panic!(
+                "Layer 0 and Layer 1 must have the same number of neurons. Found: {}, {}",
+                l0_size, l1_size
+            );
+        }
+        if o_size != 1 {
+            panic!("Layer 2 should only have one output. Found: {}", o_size);
+        }
+
+        let req_sizes = vec![i_size, l0_size, o_size].into_iter().collect();
+        let src = format!(
+            r#"
+        pub fn get_model_{name}() -> ModelData<Size{i_size}, Size{o_size}, Size{n_size}> {{
+            ModelData {{
+                norm: {norm},
+                weights: {weights},
+                fields: vec![{fields}],
+                alpha: {alpha},
+            }}
+        }}
+        "#,
+            i_size = i_size,
+            n_size = l0_size,
+            o_size = o_size,
+            name = model_name,
+            norm = self.norm.into_src(),
+            weights = self.weights.into_src(),
+            fields = fields,
+            alpha = self.alpha
+        );
+
+        (req_sizes, src)
+    }
+}
+
+#[derive(Deserialize, Debug)]
+pub struct FieldsDescribe {
+    mean: HashMap<String, f32>,
+    std: HashMap<String, f32>,
+}
+
+impl FieldsDescribe {
+    fn into_src(&self) -> String {
+        let mean = self
+            .mean
+            .iter()
+            .map(|(k, v)| format!("(\"{}\".to_string(), {})", k, v))
+            .fold(String::new(), |acc, f| acc + &f + ", ");
+        let std = self
+            .std
+            .iter()
+            .map(|(k, v)| format!("(\"{}\".to_string(), {})", k, v))
+            .fold(String::new(), |acc, f| acc + &f + ", ");
+
+        format!(
+            r#"
+        FieldsDescribe {{
+            mean: vec![{mean}].into_iter().collect(),
```

Comment:
As you noticed compile times are long.
Not a big issue, I was just wondering if they can improve a bit by not using macros inside the generated file, ie.: fiddling the file with the `vec!` expansion instead of the macro itself

Reply:
This should be fixed by the last commit, "Embed matrix data as binary strings".